src/solver/solverUtils.F90

```diff
@@ -777,18 +782,30 @@ subroutine gridVelocitiesFineLevel_block(useOldCoor, t, sps)
                    ! Determine the coordinates of the cell center,
                    ! which are stored in xc.
 
-                   xc(1) = eighth*(x(i-1,j-1,k-1,1) + x(i,j-1,k-1,1) &
-                        +         x(i-1,j,  k-1,1) + x(i,j,  k-1,1) &
-                        +         x(i-1,j-1,k,  1) + x(i,j-1,k,  1) &
-                        +         x(i-1,j,  k,  1) + x(i,j,  k,  1))
-                   xc(2) = eighth*(x(i-1,j-1,k-1,2) + x(i,j-1,k-1,2) &
-                        +         x(i-1,j,  k-1,2) + x(i,j,  k-1,2) &
-                        +         x(i-1,j-1,k,  2) + x(i,j-1,k,  2) &
-                        +         x(i-1,j,  k,  2) + x(i,j,  k,  2))
-                   xc(3) = eighth*(x(i-1,j-1,k-1,3) + x(i,j-1,k-1,3) &
-                        +         x(i-1,j,  k-1,3) + x(i,j,  k-1,3) &
-                        +         x(i-1,j-1,k,  3) + x(i,j-1,k,  3) &
-                        +         x(i-1,j,  k,  3) + x(i,j,  k,  3))
+                   xc(1) = eighth*(flowDoms(nn, groundLevel, sps)%x(i-1,j-1,k-1,1) &
```

Comment:
This is the trick that Sicheng had to do for Tapenade to run - unravel the pointers such as x =>flowDoms(nn, groundLevel, sps)%x . @eirikurj @mdolab/adflow_maintainers any comments here? We could revert back to the more elegant code with pointers at a later stage, when Tapenade can handle this

Reply:
dont we use this pointer approach in the rest of the code? and tapanade works there right? is there a difference here that makes tapenade not work with our pointers?

Reply:
@eirikurj you had a good explanation for this if I remember correctly. I guess some other parts of the code are wrapped in a different way and so the pointers are not an issue. Further wrapping these subroutines was the alternative approach to this, @SichengHe opted for unravelling the loops as "safest" approach (right?)

Reply:
@anilyil I did not recall a function contains a locally defined pointer (something like x => y(3)) can be safely differentiated. If you have something in mind, I want to take a look.

Reply:
I am not sure what you mean about locally defined pointers. I think the differentiated code in adflow is full of pointers though. Consider this metric computation example which accesses the same x array as you have above:

Master routine calls `metric_block` (https://github.com/mdolab/adflow/blob/master/src/adjoint/masterRoutines.F90#L98) after calling `setPointers` (https://github.com/mdolab/adflow/blob/master/src/adjoint/masterRoutines.F90#L94). Then `metric_block` uses the pointers imported from `blockPointers` to run its calc using the array `x` https://github.com/mdolab/adflow/blob/master/src/adjoint/adjointExtra.F90#L185. Note it does not need to go through the block data structure, it just uses the block pointer.

Now lets look at how this is differentiated: `master_d` routine calls `setPointers_d` (https://github.com/mdolab/adflow/blob/master/src/adjoint/masterRoutines.F90#L373), which is a wrapper around the regular `setPointers` but also includes the correct pointers for the memory allocated to track the derivative seeds (https://github.com/mdolab/adflow/blob/master/src/utils/utils.F90#L3584-L3595). After `master_d` calls `setPointers_d`, it can then call `metric_block_d` routine (https://github.com/mdolab/adflow/blob/master/src/adjoint/masterRoutines.F90#L387), which is AD'd using the block pointers: https://github.com/mdolab/adflow/blob/master/src/adjoint/outputForward/adjointextra_d.f90#L400 

This way, the code still operates on block pointers and uses tapenade to differentiate against these block pointers. We manually handle the block pointers used for the derivative seeds by calling `setPointers_d`. I believe the AD pointers are also manually allocated by us in: https://github.com/mdolab/adflow/blob/master/src/modules/block.F90#L783-L785 I only used fwd AD examples here but same logic holds for reverse AD as well.

With all of this info in mind, I don't understand what is limiting the use of block pointers here. There already is a call to `setPointers` right before calling `gridVelocitiesFineLevel_block` in `solverUtils` in the current PR: https://github.com/mdolab/adflow/pull/121/files/db808880b542f45a484365090ce6c97f7e8373df#diff-c30a13783d79386e711f67eef5ddf3d84d06b73e8507b19b238d335e43dd1f0eL387 What we would need to do is to modify the `gridVelocitiesFineLevel_block` to operate on block pointers, and then manually differentiate `gridVelocitiesFineLevel` to have a call to `setPointers_d` before calling `gridVelocitiesFineLevel_block_d`. 

I may be totally missing something but this was my understanding of how ADflow is differentiated using whatever tapenade version w/ or w/o support for pointers.

Reply:
I haven't started reading this yet but I assume this would fit in the docs :wink:

Reply:
@anilyil  Yep, I got it that we can do AD for pointers. However, if you define y=>x where x is some array, then Tapenade struggles to differentiate y... That is what I mean by the locally defined pointer. I am not sure whether that is accurate nomenclature. @eirikurj may be able to explain this in more detail.

Reply:
Okay, I get what you mean by the locally defined pointer. Is this the problem here though? For example, do you need `flowdoms(...)%x` instead of just importing `x` from `blockPointers`?  Some other pointer stuff that might be in the code that I remember were related to faces, and those can be imported from BC data I believe, instead of a locally defined pointer to a slice of block data.